File: src/visual-graph/prune/node-matches-query.ts
import type { ParsedRootQuery } from "../../cli/root-query/parsed-root-query.js";
import type { VisualNode } from "../model.js";
import { NAME_QUERY_EXCLUDED } from "./name-query-excluded.js";

export function nodeMatchesQuery(
  node: VisualNode,
  q: ParsedRootQuery,
): boolean {
  const startLine = node.line;
  const endLine = node.endLine ?? node.line;
  switch (q.kind) {
    case "line":
      return q.line >= startLine && q.line <= endLine;
    case "line-name":
      return q.line >= startLine && q.line <= endLine && node.name === q.name;
    case "range":
      return startLine <= q.end && endLine >= q.start;
    case "range-name":
      return startLine <= q.end && endLine >= q.start && node.name === q.name;
    case "name":
      return !NAME_QUERY_EXCLUDED.has(node.kind) && node.name === q.name;
    case "line-or-name":
      // resolveAmbiguousQueries rewrites every line-or-name into Line or
      // Name before pruning runs, so this branch is unreachable. The arm
      // exists to keep the switch exhaustive.
      return false;
  }
}

File: crates/unsnarl-visual-graph/src/prune/node_matches_query.rs
//! Per-node matcher for a single `-r` query.

use unsnarl_ir::SourceLine;
use unsnarl_root_query::ParsedRootQuery;

use crate::prune::name_query_excluded::is_name_query_excluded;
use crate::visual_node::VisualNode;

pub fn node_matches_query(node: &VisualNode, q: &ParsedRootQuery) -> bool {
    let start_line = SourceLine(node.line());
    let end_line = SourceLine(node.end_line().unwrap_or_else(|| node.line()));
    match q {
        ParsedRootQuery::Line { line, .. } => *line >= start_line && *line <= end_line,
        ParsedRootQuery::LineName { line, name, .. } => {
            *line >= start_line && *line <= end_line && node.name() == name.as_str()
        }
        ParsedRootQuery::Range { start, end, .. } => start_line <= *end && end_line >= *start,
        ParsedRootQuery::RangeName {
            start, end, name, ..
        } => start_line <= *end && end_line >= *start && node.name() == name.as_str(),
        ParsedRootQuery::Name { name, .. } => {
            !is_name_query_excluded(node.kind()) && node.name() == name.as_str()
        }
        ParsedRootQuery::LineOrName { .. } => {
            // resolveAmbiguousQueries rewrites every line-or-name into
            // Line or Name before pruning runs, so this branch is
            // unreachable. The arm exists to keep the match exhaustive.
            false
        }
    }
}

#[cfg(test)]
#[path = "node_matches_query_test.rs"]
mod node_matches_query_test;
